src/Workspaces/Core/Portable/Rename/Renamer.RenameSymbolDocumentAction.cs

```diff
@@ -0,0 +1,115 @@
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Rename
+{
+    public static partial class Renamer
+    {
+        /// <summary>
+        /// Action that will rename a type to match the current document name
+        /// </summary>
+        internal class RenameSymbolDocumentAction : RenameDocumentAction
+        {
+            private readonly AnalysisResult _analysis;
+
+            private RenameSymbolDocumentAction(
+                AnalysisResult analysis,
+                OptionSet optionSet,
+                ImmutableArray<ErrorResource> errors)
+                : base(errors, optionSet)
+            {
+                _analysis = analysis;
+            }
+
+            public override string GetDescription(CultureInfo? culture)
+                => string.Format(WorkspacesResources.ResourceManager.GetString("Rename_0_to_1", culture ?? WorkspacesResources.Culture)!, _analysis.OriginalDocumentName, _analysis.NewDocumentName);
+
+            internal override async Task<Solution> GetModifiedSolutionAsync(Document document, CancellationToken cancellationToken)
+            {
+                var solution = document.Project.Solution;
+                var matchingTypeDeclaration = await GetMatchingTypeDeclarationAsync(document, _analysis.OriginalSymbolName!, cancellationToken).ConfigureAwait(false);
+
+                if (matchingTypeDeclaration is object)
+                {
+                    var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                    var symbol = semanticModel.GetDeclaredSymbol(matchingTypeDeclaration, cancellationToken);
+
+                    solution = await RenameSymbolAsync(solution, symbol, _analysis.NewSymbolName, OptionSet, cancellationToken).ConfigureAwait(false);
+                }
+
+                return solution;
+            }
+
+            private static async Task<SyntaxNode?> GetMatchingTypeDeclarationAsync(Document document, string name, CancellationToken cancellationToken)
+            {
+                var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+                var typeDeclarations = syntaxRoot.DescendantNodesAndSelf().Where(syntaxFacts.IsTypeDeclaration);
+                return typeDeclarations.FirstOrDefault(d => syntaxFacts.GetDisplayName(d, DisplayNameOptions.None).Equals(name, StringComparison.OrdinalIgnoreCase));
+            }
+
+            public static async Task<RenameSymbolDocumentAction?> TryCreateAsync(Document document, string newName, OptionSet optionSet, CancellationToken cancellationToken)
+            {
+                var analysis = await AnalysisResult.CreateAsync(document, newName, optionSet, cancellationToken).ConfigureAwait(false);
+
+                if (analysis.ShouldApplyAction)
```

Comment:
why is this in the anlaysis result.  why not have CreateAsync check this and just return null if it shouldn't be applied?

Reply:
Analysis happens either way right? It seems more strange that we take time to analyze then return nothing. 

Maybe renaming this to `SymbolNeedsRename` would be better suited. That way it's the analysis saying that the symbol needs rename, and the action creation determining that if a symbol rename is needed then we need to create an action to do so.

Reply:
i just don't get it.  we package up the info to put into an object to pass to the caller, who has to then check that objct to decide not to do anything.  

I would jsut expect hte analysis to all be in one location, the result to be smaller (and just effectively data), and if we can't do it, we just pass back a null result as that totally indicates that anlaysis didn't work.

Reply:
Analysis is all in one location. What I'm saying is this: 

Analysis provides data about specific action. In this case, that data could be information that the symbol needs to be renamed. That's a valid analysis to say it doesn't need to be renamed. Why would return null there? 

That said, we're really nitpicking over a small private class design. If this is blocking I'll change it because I don't feel strongly enough to argue over 2 lines of code that do the same thing.

Reply:
I'm agreeing with @CyrusNajmabadi here for one other reason: OriginalSymbolName seems like it is marked as nullable to allow for how ShouldApplyAction, which then results in a random ! being written earlier. The null annotation has to encompass both the valid and invalid states; if this just returned null then it's just having to deal with the valid states, which makes the rest of the code easier to validate.

If we had discriminated unions then you'd have a Pass (with all of the state) and a Fail with no state, and it'd be clearer that way, but alas we don't have those...yet...

Reply:
Fixed